experimental/execution/ptx-executor/src/finalizer.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{common::TxnIdx, metrics::TIMER, state_view::OverlayedStateView};
+use aptos_logger::trace;
+use aptos_metrics_core::TimerHelper;
+use aptos_state_view::StateView;
+use aptos_types::{transaction::TransactionOutput, write_set::TransactionWrite};
+use aptos_vm_types::output::VMOutput;
+use rayon::Scope;
+use std::{
+    collections::VecDeque,
+    sync::mpsc::{channel, Sender},
+};
+
+pub(crate) struct PtxFinalizer;
+
+impl PtxFinalizer {
+    pub fn spawn<'scope, 'view: 'scope>(
+        scope: &Scope<'scope>,
+        base_view: &'view (dyn StateView + Sync),
+        result_tx: Sender<TransactionOutput>,
+    ) -> PtxFinalizerClient {
+        let (work_tx, work_rx) = channel();
+        scope.spawn(move |_scope| {
+            let _timer = TIMER.timer_with(&["finalizer_block_total"]);
+            let mut worker = Worker::new(base_view, result_tx);
+            loop {
+                match work_rx.recv().expect("Channel closed.") {
+                    Command::AddVMOutput { txn_idx, vm_output } => {
+                        worker.add_vm_output(txn_idx, vm_output)
+                    },
+                    Command::FinishBlock => {
+                        worker.finish_block();
+                        break;
+                    },
+                }
+            }
+        });
+        PtxFinalizerClient { work_tx }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct PtxFinalizerClient {
+    work_tx: Sender<Command>,
+}
+
+impl PtxFinalizerClient {
+    pub fn add_vm_output(&self, txn_idx: TxnIdx, vm_output: VMOutput) {
+        self.send_to_worker(Command::AddVMOutput { txn_idx, vm_output })
+    }
+
+    pub fn finish_block(&self) {
+        self.send_to_worker(Command::FinishBlock)
+    }
+
+    fn send_to_worker(&self, command: Command) {
+        self.work_tx.send(command).expect("Work thread died.");
+    }
+}
+
+struct Worker<'view> {
+    result_tx: Sender<TransactionOutput>,
+    vm_output_buffer: VecDeque<Option<VMOutput>>,
+    next_idx: TxnIdx,
+    state_view: OverlayedStateView<'view>,
+}
+
+impl<'view> Worker<'view> {
+    fn new(base_view: &'view (dyn StateView + Sync), result_tx: Sender<TransactionOutput>) -> Self {
+        Self {
+            result_tx,
+            vm_output_buffer: VecDeque::new(),
+            next_idx: 0,
+            state_view: OverlayedStateView::new(base_view),
+        }
+    }
+
+    fn add_vm_output(&mut self, txn_idx: TxnIdx, vm_output: VMOutput) {
+        trace!("seen txn: {}", txn_idx);
+        assert!(txn_idx >= self.next_idx);
+        let idx_in_buffer = txn_idx - self.next_idx;
+        if self.vm_output_buffer.len() < idx_in_buffer + 1 {
+            self.vm_output_buffer.resize(idx_in_buffer + 1, None);
```

Comment:
Too fancy, and might actually worse than allocating the whole space at the beginning.

Reply:
I'll add `with_capacity()` on line 76